src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -248,52 +292,174 @@ public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int b
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = s_avxDecodeShuffleVec;
+            Vector256<int> permuteVec = s_avxDecodePermuteVec;
+
+            //while (remaining >= 45)
+            do
+            {
+                AssertRead<Vector256<sbyte>>(ref src, ref srcStart, sourceLength);
+                Vector256<sbyte> str = Unsafe.As<byte, Vector256<sbyte>>(ref src);
+
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLo, loNibbles);
+                Vector256<sbyte> zero = Vector256<sbyte>.Zero;
+
+                // https://github.com/dotnet/coreclr/issues/21247
+                if (Avx2.MoveMask(Avx2.CompareGreaterThan(Avx2.And(lo, hi), zero)) != 0)
+                    break;
+
+                Vector256<sbyte> eq2F = Avx2.CompareEqual(str, mask2F);
+                Vector256<sbyte> shift = Avx2.Shuffle(lutShift, Avx2.Add(eq2F, hiNibbles));
+                str = Avx2.Add(str, shift);
+
+                Vector256<short> merge_ab_and_bc = Avx2.MultiplyAddAdjacent(str.AsByte(), shuffleConstant0);
+                Vector256<int> @out = Avx2.MultiplyAddAdjacent(merge_ab_and_bc, shuffleConstant1);
+                @out = Avx2.Shuffle(@out.AsSByte(), shuffleVec).AsInt32();
+                str = Avx2.PermuteVar8x32(@out, permuteVec).AsSByte();
+
+                AssertWrite<Vector256<sbyte>>(ref destBytes, ref destStart, destLength);
+                // As has better CQ than WriteUnaligned
+                // https://github.com/dotnet/coreclr/issues/21132
+                Unsafe.As<byte, Vector256<sbyte>>(ref destBytes) = str;
+
+                src = ref Unsafe.Add(ref src, 32);
+                destBytes = ref Unsafe.Add(ref destBytes, 24);
+            }
+            while (Unsafe.IsAddressLessThan(ref src, ref simdSrcEnd));
+
+            // Cast to ulong to avoid the overflow-check. Codegen for x86 is still good.
+            sourceIndex = (uint)(ulong)Unsafe.ByteOffset(ref srcStart, ref src);
+            destIndex = (uint)(ulong)Unsafe.ByteOffset(ref destStart, ref destBytes);
+
+            src = ref srcStart;
+            destBytes = ref destStart;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Ssse3Decode(ref byte src, ref byte destBytes, int sourceLength, int destLength, ref uint sourceIndex, ref uint destIndex)
+        {
+            ref byte srcStart = ref src;
+            ref byte destStart = ref destBytes;
+            ref byte simdSrcEnd = ref Unsafe.Add(ref src, (IntPtr)((uint)sourceLength - 24 + 1));
+
+            // The JIT won't hoist these "constants", so help him
+            Vector128<sbyte> lutHi = s_sseDecodeLutHi;
```

Comment:
how many of these are actually kept in register rather than being spilled/unspilled?

Reply:
In the dasm for all loops no spills occur, so everything is kept in registers.
But I don't have a pure SSE machine to test, and with `COMPlus_EnableAVX2=0` just AVX2 is disabled, but the registers get still used.

Reply:
Good to know how to disable VEX 👍 

#### Encoding

No spills.

```asm
G_M39785_IG06:
       440F1006             movups   xmm8, xmmword ptr [rsi]
       66440F3800C0         pshufb   xmm8, xmm0
       450F28C8             movaps   xmm9, xmm8
       66440FDBC9           pand     xmm9, xmm1
       66440FDBC2           pand     xmm8, xmm2
       66440FE4CB           pmulhuw  xmm9, xmm3
       66440FD5C4           pmullw   xmm8, xmm4
       66450FEBC8           por      xmm9, xmm8
       450F28C1             movaps   xmm8, xmm9
       450F28C8             movaps   xmm9, xmm8
       66440FD8CD           psubusb  xmm9, xmm5
       450F28D0             movaps   xmm10, xmm8
       66440F64D6           pcmpgtb  xmm10, xmm6
       66450FF8CA           psubb    xmm9, xmm10
       440F28D7             movaps   xmm10, xmm7
       66450F3800D1         pshufb   xmm10, xmm9
       66450FFCC2           paddb    xmm8, xmm10
       440F1102             movups   xmmword ptr [rdx], xmm8
       4883C60C             add      rsi, 12
       4883C210             add      rdx, 16
       483BF7               cmp      rsi, rdi
       729E                 jb       SHORT G_M39785_IG06
```

#### Decoding 

No spills.

```asm
G_M25135_IG05:
       0F103F               movups   xmm7, xmmword ptr [rdi]
       440F28C7             movaps   xmm8, xmm7
       66410F72D004         psrld    xmm8, 4
       66440FDBC3           pand     xmm8, xmm3
       440F28CF             movaps   xmm9, xmm7
       66440FDBCB           pand     xmm9, xmm3
       440F28D0             movaps   xmm10, xmm0
       66450F3800D0         pshufb   xmm10, xmm8
       450F57DB             xorps    xmm11, xmm11
       440F28E1             movaps   xmm12, xmm1
       66450F3800E1         pshufb   xmm12, xmm9
       450F28CC             movaps   xmm9, xmm12
       66450FDBCA           pand     xmm9, xmm10
       66450F64CB           pcmpgtb  xmm9, xmm11
       66410FD7D1           pmovmskb  edx, xmm9
       85D2                 test     edx, edx
       7540                 jne      SHORT G_M25135_IG06
       440F28CF             movaps   xmm9, xmm7
       66440F74CB           pcmpeqb  xmm9, xmm3
       66450FFCC8           paddb    xmm9, xmm8
       440F28C2             movaps   xmm8, xmm2
       66450F3800C1         pshufb   xmm8, xmm9
       66410FFCF8           paddb    xmm7, xmm8
       660F3804FC           pmaddubsw xmm7, xmm4
       660FF5FD             pmaddwd  xmm7, xmm5
       660F3800FE           pshufb   xmm7, xmm6
       0F113E               movups   xmmword ptr [rsi], xmm7
       4883C710             add      rdi, 16
       4883C60C             add      rsi, 12
       483BF8               cmp      rdi, rax
       0F8277FFFFFF         jb       G_M25135_IG05
```